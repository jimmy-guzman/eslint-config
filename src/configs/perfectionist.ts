import type { ESLint } from "eslint";

import perfectionist from "eslint-plugin-perfectionist";

import type { TypedConfigItem } from "../types";

import { perfectionistRules } from "../rules/perfectionist";

export const perfectionistConfig = () => {
  return [
    {
      name: "jimmy.codes/perfectionist",
      plugins: {
        // TODO: remove unknown conversion
        perfectionist: perfectionist as unknown as ESLint.Plugin,
      },
      rules: perfectionistRules,
    },
  ] satisfies TypedConfigItem[];
};
