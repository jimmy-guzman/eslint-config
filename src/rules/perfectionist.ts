import perfectionist from "eslint-plugin-perfectionist";

import { type Rules } from "../types";

export const perfectionistRules = {
  ...perfectionist.configs["recommended-natural"].rules,
  "perfectionist/sort-imports": [
    "error",
    {
      customGroups: { type: {}, value: {} },
      environment: "node",
      groups: [
        "builtin",
        "type",
        "external",
        "internal-type",
        "internal",
        ["parent-type", "sibling-type", "index-type"],
        ["parent", "sibling", "index"],
        "object",
        "unknown",
      ],
      internalPattern: ["~/**", "@/**"],
      order: "asc",
      type: "natural",
    },
  ],
} satisfies Rules;
