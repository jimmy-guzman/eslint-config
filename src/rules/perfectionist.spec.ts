import { perfectionistRules } from "./perfectionist";

test("should create perfectionist rules", () => {
  expect(perfectionistRules).toMatchSnapshot();
});
