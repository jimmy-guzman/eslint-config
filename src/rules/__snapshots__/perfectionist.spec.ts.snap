// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should create perfectionist rules 1`] = `
{
  "perfectionist/sort-array-includes": [
    "error",
    {
      "order": "asc",
      "type": "natural",
    },
  ],
  "perfectionist/sort-classes": [
    "error",
    {
      "order": "asc",
      "type": "natural",
    },
  ],
  "perfectionist/sort-decorators": [
    "error",
    {
      "order": "asc",
      "type": "natural",
    },
  ],
  "perfectionist/sort-enums": [
    "error",
    {
      "order": "asc",
      "type": "natural",
    },
  ],
  "perfectionist/sort-exports": [
    "error",
    {
      "order": "asc",
      "type": "natural",
    },
  ],
  "perfectionist/sort-heritage-clauses": [
    "error",
    {
      "order": "asc",
      "type": "natural",
    },
  ],
  "perfectionist/sort-imports": [
    "error",
    {
      "customGroups": {
        "type": {},
        "value": {},
      },
      "environment": "node",
      "groups": [
        "side-effect-style",
        "builtin",
        "type",
        "external",
        "internal-type",
        "internal",
        [
          "parent-type",
          "sibling-type",
          "index-type",
        ],
        [
          "parent",
          "sibling",
          "index",
        ],
        "object",
        "style",
        "unknown",
      ],
      "internalPattern": [
        "^~/.*",
        "^@/.*",
      ],
      "order": "asc",
      "type": "natural",
    },
  ],
  "perfectionist/sort-interfaces": [
    "error",
    {
      "order": "asc",
      "type": "natural",
    },
  ],
  "perfectionist/sort-intersection-types": [
    "error",
    {
      "order": "asc",
      "type": "natural",
    },
  ],
  "perfectionist/sort-jsx-props": [
    "error",
    {
      "order": "asc",
      "type": "natural",
    },
  ],
  "perfectionist/sort-maps": [
    "error",
    {
      "order": "asc",
      "type": "natural",
    },
  ],
  "perfectionist/sort-modules": "off",
  "perfectionist/sort-named-exports": [
    "error",
    {
      "order": "asc",
      "type": "natural",
    },
  ],
  "perfectionist/sort-named-imports": [
    "error",
    {
      "order": "asc",
      "type": "natural",
    },
  ],
  "perfectionist/sort-object-types": [
    "error",
    {
      "order": "asc",
      "type": "natural",
    },
  ],
  "perfectionist/sort-objects": [
    "error",
    {
      "order": "asc",
      "type": "natural",
    },
  ],
  "perfectionist/sort-sets": [
    "error",
    {
      "order": "asc",
      "type": "natural",
    },
  ],
  "perfectionist/sort-switch-case": [
    "error",
    {
      "order": "asc",
      "type": "natural",
    },
  ],
  "perfectionist/sort-union-types": [
    "error",
    {
      "order": "asc",
      "type": "natural",
    },
  ],
  "perfectionist/sort-variable-declarations": [
    "error",
    {
      "order": "asc",
      "type": "natural",
    },
  ],
}
`;
